fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -359,30 +335,56 @@ public void close() {
-                                                   @Nullable AtomicLong recordsScanned) {
-        if (increaseLimitAfter > 0) {
-            if (exception == null) {
-                successCount++;
-                if (successCount >= increaseLimitAfter && limit < maxLimit) {
-                    increaseLimit();
-                }
-            } else {
-                successCount = 0;
-            }
-        }
-        if (recordsScanned != null) {
-            if (exception == null) {
-                totalRecordsScanned.addAndGet(recordsScanned.get());
+    interface ExceptionHandler<R> {
+        CompletableFuture<Boolean> apply(CompletableFuture<R> ret, AtomicLong toWait, Throwable e, @Nullable List<Object> additionalLogMessageKeyValues);
+    }
+
+    private <R> CompletableFuture<Boolean> completeExceptionally(CompletableFuture<R> ret, Throwable e) {
+        ret.completeExceptionally(runner.getDatabase().mapAsyncToSyncException(e));
+        return AsyncUtil.READY_FALSE;
+    }
+
+    @VisibleForTesting
+    <R> CompletableFuture<R> buildAsync(@Nonnull BiFunction<FDBRecordStore, AtomicLong, CompletableFuture<R>> buildFunction,
+                                        boolean limitControl,
+                                        @Nullable List<Object> additionalLogMessageKeyValues) {
+        AtomicLong recordsScanned = new AtomicLong(0);
+        return runAsync(store -> buildFunction.apply(store, recordsScanned),
+                // Run after a single transactional call within runAsync.
+                (result, exception) -> {
+                    if (limitControl) {
+                        tryToIncreaseLimit(exception);
+                    }
+                    // Update records scanned.
+                    if (exception == null) {
+                        totalRecordsScanned.addAndGet(recordsScanned.get());
+                    } else {
+                        recordsScanned.set(0);
+                    }
+                    return Pair.of(result, exception);
+                },
+                limitControl ? this::tryToDecreaseLimit : null,
+                additionalLogMessageKeyValues
+        );
+    }
+
+    @VisibleForTesting
+    <R> CompletableFuture<Boolean> tryToDecreaseLimit(CompletableFuture<R> ret, AtomicLong toWait, Throwable e, @Nullable List<Object> additionalLogMessageKeyValues) {
+        FDBException fdbE = getFDBException(e);
+        if (fdbE == null) {
+            return completeExceptionally(ret, e);
+        } else {
+            if (lessenWorkCodes.contains(fdbE.getCode())) {
+                decreaseLimit(fdbE, additionalLogMessageKeyValues);
+                long delay = (long)(Math.random() * toWait.get());
+                toWait.set(Math.min(delay * 2, FDBDatabaseFactory.instance().getMaxDelayMillis()));
+                return MoreAsyncUtil.delayedFuture(delay, TimeUnit.MILLISECONDS).thenApply(vignore3 -> true);
```

Comment:
So in the case of `buildEndpoints`, it will no longer have the delay if it gets a `lessenWorkCodes`, right? I don't think that's what we want, I think we still want that, just not decreasing the limit.